lib/cache/access_list.go

```diff
@@ -0,0 +1,399 @@
+			upstreamRead = true
+			return c.Config.AccessLists.GetAccessList(ctx, s)
+		},
+		clone: func(al *accesslist.AccessList) *accesslist.AccessList {
+			return al.Clone()
+		},
+	}
+	out, err := getter.get(ctx, name)
+	if trace.IsNotFound(err) && !upstreamRead {
+		// fallback is sane because method is never used
+		// in construction of derivative caches.
+		if item, err := c.Config.AccessLists.GetAccessList(ctx, name); err == nil {
+			return item, nil
+		}
+	}
+	return out, trace.Wrap(err)
+}
+
+const (
+	accessListMemberStoreNameIndex = "name"
+	accessListMemberStoreKindIndex = "kind"
+)
+
+func newAccessListMemberCollection(upstream services.AccessLists, w types.WatchKind) (*collection[*accesslist.AccessListMember], error) {
+	if upstream == nil {
+		return nil, trace.BadParameter("missing parameter AccessLists")
+	}
+
+	return &collection[*accesslist.AccessListMember]{
+		store: newStore(map[string]func(*accesslist.AccessListMember) string{
+			accessListMemberStoreNameIndex: func(r *accesslist.AccessListMember) string {
+				return r.Spec.AccessList + "/" + r.GetName()
+			},
+			accessListMemberStoreKindIndex: func(r *accesslist.AccessListMember) string {
+				return r.Spec.AccessList + "/" + r.Spec.MembershipKind + "/" + r.GetName()
+			},
+		}),
+		fetcher: func(ctx context.Context, loadSecrets bool) ([]*accesslist.AccessListMember, error) {
+			var resources []*accesslist.AccessListMember
+			var nextToken string
+			for {
+				var page []*accesslist.AccessListMember
+				var err error
+				page, nextToken, err = upstream.ListAllAccessListMembers(ctx, 0 /* page size */, nextToken)
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+
+				resources = append(resources, page...)
+
+				if nextToken == "" {
+					break
+				}
+			}
+			return resources, nil
+		},
+		headerTransform: func(hdr *types.ResourceHeader) *accesslist.AccessListMember {
+			return &accesslist.AccessListMember{
+				ResourceHeader: header.ResourceHeader{
+					Kind:    hdr.Kind,
+					Version: hdr.Version,
+					Metadata: header.Metadata{
+						Name: hdr.Metadata.Name,
+					},
+				},
+				Spec: accesslist.AccessListMemberSpec{
+					AccessList: hdr.Metadata.Description,
+				},
+			}
+		},
+		watch: w,
+	}, nil
+}
+
+// CountAccessListMembers will count all access list members.
+func (c *Cache) CountAccessListMembers(ctx context.Context, accessListName string) (uint32, uint32, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/CountAccessListMembers")
+	defer span.End()
+
+	rg, err := acquireReadGuard(c, c.collections.accessListMembers)
+	if err != nil {
+		return 0, 0, trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if !rg.ReadCache() {
+		count, listCount, err := c.Config.AccessLists.CountAccessListMembers(ctx, accessListName)
+		return count, listCount, trace.Wrap(err)
+	}
+
+	startKey := accessListName + "/" + accesslist.MembershipKindList
+	endKey := sortcache.NextKey(startKey)
```

Comment:
Safeguard against the addition of new `accesslist.MembershipKind` variants with names that share a prefix

```suggestion
	startKey := accessListName + "/" + accesslist.MembershipKindList + "/"
	endKey := sortcache.NextKey(startKey)
```